src/shogun/lib/SGMatrix.h

```diff
@@ -41,7 +41,244 @@ template<class T> class SGMatrix : public SGReferencedData
+			}
+			column_iterator& operator--()
+			{
+				--m_col_idx;
+				return *this;
+			}
+			column_iterator operator--(int)
+			{
+				column_iterator retval(*this);
+				--(*this);
+				return retval;
+			}
+
+			bool operator==(const column_iterator& other) const
+			{
+				return m_mat == other.m_mat && m_col_idx == other.m_col_idx;
+			}
+			bool operator!=(const column_iterator& other) const
+			{
+				return !(*this == other);
+			}
+
+			value_type operator*()
+			{
+				return m_mat.get_column(m_col_idx);
+			}
+
+			column_iterator& operator+=(difference_type d)
+			{
+				m_col_idx += d;
+				return *this;
+			}
+			column_iterator& operator-=(difference_type d)
+			{
+				m_col_idx -= d;
+				return *this;
+			}
+
+			column_iterator operator+(difference_type d) const
+			{
+				return column_iterator(m_mat, m_col_idx + d);
+			}
+			column_iterator operator-(difference_type d) const
+			{
+				return column_iterator(m_mat, m_col_idx - d);
+			}
+
+			value_type operator[](difference_type d) const
+			{
+				return *column_iterator(m_mat, m_col_idx + d);
+			}
+
+			bool operator<(const column_iterator& other) const
+			{
+				return m_col_idx < other.m_col_idx;
+			}
+			bool operator>(const column_iterator& other) const
+			{
+				return m_col_idx > other.m_col_idx;
+			}
+			bool operator<=(const column_iterator& other) const
+			{
+				return m_col_idx <= other.m_col_idx;
+			}
+			bool operator>=(const column_iterator& other) const
+			{
+				return m_col_idx >= other.m_col_idx;
+			}
+
+			difference_type operator-(const column_iterator& other) const
+			{
+				return m_col_idx - other.m_col_idx;
+			}
+
+		private:
+			SGMatrix<T>& m_mat;
+			difference_type m_col_idx;
+		};
+
+		class const_column_iterator
+		{
+		public:
+			// Iterator traits
+			using difference_type = index_t;
+			using value_type = const SGVector<T>;
+			using pointer = const SGVector<T>*;
+			using reference = const SGVector<T>&;
+			using iterator_category = std::random_access_iterator_tag;
+
+			const_column_iterator(const SGMatrix<T>& mat, index_t idx = 0)
+			    : m_mat(mat), m_col_idx(idx)
+			{
+			}
+			const_column_iterator(const const_column_iterator& other)
+			    : m_mat(other.m_mat), m_col_idx(other.m_col_idx)
+			{
+			}
+
+			const_column_iterator& operator++()
+			{
+				++m_col_idx;
+				return *this;
+			}
+			const_column_iterator operator++(int)
+			{
+				const_column_iterator retval(*this);
+				++(*this);
+				return retval;
+			}
+			const_column_iterator& operator--()
+			{
+				--m_col_idx;
+				return *this;
+			}
+			const_column_iterator operator--(int)
+			{
+				const_column_iterator retval(*this);
+				--(*this);
+				return retval;
+			}
 
+			bool operator==(const const_column_iterator& other) const
+			{
+				return m_mat == other.m_mat && m_col_idx == other.m_col_idx;
+			}
+			bool operator!=(const const_column_iterator& other) const
+			{
+				return !(*this == other);
+			}
+
+			value_type operator*()
```

Comment:
Here you can return reference

Reply:
Correct me if I'm wrong but m_mat.get_column(m_col_idx) returns an SGVector<T>, so it's a local variable. We can't return reference of local variables in a function right? (We can but that involves using value_type).

Reply:
Ah yes, sorry